File: sdks/python/src/zro_sdk/module.py
"""Module system for ZRO backend applications.

Modules are self-contained units that contribute commands, event handlers,
and lifecycle hooks to a ZRO app. They declare dependencies, are resolved
in topological order, and have optional init/destroy lifecycle hooks.

Example::

    from zro_sdk import ZroModule, ModuleMeta, ModuleRegistrar, AppContext

    class GreetModule(ZroModule):
        @property
        def meta(self) -> ModuleMeta:
            return ModuleMeta(name="greet", version="0.1.0")

        def register(self, r: ModuleRegistrar) -> None:
            @r.command("greet")
            async def greet(ctx: AppContext, name: str = "world") -> str:
                return f"Hello, {name}!"

    # Usage:
    app = ZroApp()
    app.module(GreetModule())
    app.run()
"""

from __future__ import annotations

import abc
from collections import deque
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Callable, Coroutine, Optional

from .context import AppContext

CommandHandler = Callable[..., Coroutine[Any, Any, Any]]
EventHandler = Callable[..., Coroutine[Any, Any, None]]
LifecycleHandler = Callable[[AppContext], Coroutine[Any, Any, None]]
InitHook = Callable[["ModuleInitContext"], Coroutine[Any, Any, None]]
DestroyHook = Callable[[], Coroutine[Any, Any, None]]


# ── Module Metadata ──────────────────────────────────────────────

@dataclass
class ModuleMeta:
    """Metadata describing a module: identity, version, and dependencies."""

    name: str
    version: str = "0.1.0"
    description: Optional[str] = None
    dependencies: list[str] = field(default_factory=list)


# ── Module Init Context ─────────────────────────────────────────

@dataclass
class ModuleInitContext:
    """Context available during module initialization (after IPC handshake)."""

    slug: str
    data_dir: Path


# ── Module Registrar ─────────────────────────────────────────────

class ModuleRegistrar:
    """Builder passed to :meth:`ZroModule.register` for contributing handlers.

    Mirrors the ``ZroApp`` decorator API so modules register in the same way.
    """

    def __init__(self) -> None:
        self.commands: dict[str, CommandHandler] = {}
        self.event_handlers: dict[str, EventHandler] = {}
        self.lifecycle_handlers: dict[str, LifecycleHandler] = {}
        self.init_hooks: list[InitHook] = []
        self.destroy_hooks: list[DestroyHook] = []

    def command(self, name: str):
        """Register a command handler (WS invoke + HTTP API).

        Usage::

            @r.command("greet")
            async def greet(ctx: AppContext, name: str) -> str:
                return f"Hello, {name}!"
        """

        def decorator(func: CommandHandler):
            self.commands[name] = func
            return func

        return decorator

    def on_event(self, event: str):
        """Register a WS event handler (fire-and-forget).

        Usage::

            @r.on_event("term:input")
            async def handle_input(ctx: AppContext, data: str):
                pass
        """

        def decorator(func: EventHandler):
            self.event_handlers[event] = func
            return func

        return decorator

    def on(self, event: str):
        """Register a lifecycle handler (``client:connected``, etc.).

        Usage::

            @r.on("client:connected")
            async def on_connect(ctx: AppContext):
                pass
        """

        def decorator(func: LifecycleHandler):
            self.lifecycle_handlers[event] = func
            return func

        return decorator

    def on_init(self, handler: InitHook) -> InitHook:
        """Register an init hook, called after IPC handshake.

        Usage::

            @r.on_init
            async def init(ctx: ModuleInitContext):
                pass
        """
        self.init_hooks.append(handler)
        return handler

    def on_destroy(self, handler: DestroyHook) -> DestroyHook:
        """Register a destroy hook, called during shutdown (reverse order).

        Usage::

            @r.on_destroy
            async def cleanup():
                pass
        """
        self.destroy_hooks.append(handler)
        return handler


# ── Module Base Class ────────────────────────────────────────────

class ZroModule(abc.ABC):
    """A ZRO backend module. Subclass this to package reusable
    commands, event handlers, and lifecycle hooks."""

    @property
    @abc.abstractmethod
    def meta(self) -> ModuleMeta:
        """Module metadata (name, version, dependencies)."""
        ...

    @abc.abstractmethod
    def register(self, registrar: ModuleRegistrar) -> None:
        """Register handlers on the provided registrar.

        Called once during app setup, in dependency order.
        """
        ...


# ── Dependency Resolution ────────────────────────────────────────

def resolve_module_order(modules: list[ZroModule]) -> list[int]:
    """Resolve module initialization order via topological sort.

    Returns indices into *modules* in the order they should be initialized.

    Raises:
        ValueError: on missing dependencies or circular references.
    """
    name_to_idx: dict[str, int] = {}
    for i, m in enumerate(modules):
        name_to_idx[m.meta.name] = i

    n = len(modules)
    in_degree = [0] * n
    adj: list[list[int]] = [[] for _ in range(n)]

    for i, m in enumerate(modules):
        for dep in m.meta.dependencies:
            dep_idx = name_to_idx.get(dep)
            if dep_idx is None:
                raise ValueError(
                    f"Module '{m.meta.name}' depends on '{dep}' which is not registered"
                )
            adj[dep_idx].append(i)
            in_degree[i] += 1

    # Kahn's algorithm
    queue = deque(i for i in range(n) if in_degree[i] == 0)
    order: list[int] = []

    while queue:
        node = queue.popleft()
        order.append(node)
        for nxt in adj[node]:
            in_degree[nxt] -= 1
            if in_degree[nxt] == 0:
                queue.append(nxt)

    if len(order) != n:
        raise ValueError("Circular dependency detected among modules")

    return order

File: sdks/nodejs/dist-test/src/module.js
"use strict";
/**
 * Module system for ZRO backend applications (Node.js).
 *
 * Modules are self-contained units that contribute commands, event handlers,
 * and lifecycle hooks to a ZRO app. They declare dependencies, are resolved
 * in topological order, and have optional init/destroy lifecycle hooks.
 *
 * @example
 * ```ts
 * import { ZroModule, ModuleMeta, ModuleRegistrar } from '@zro/sdk';
 *
 * const greetModule: ZroModule = {
 *     meta: { name: 'greet', version: '0.1.0' },
 *     register(r: ModuleRegistrar) {
 *         r.command('greet', async (ctx, params) => {
 *             return `Hello, ${params.name ?? 'world'}!`;
 *         });
 *     },
 * };
 *
 * const app = new ZroApp();
 * app.module(greetModule);
 * app.run();
 * ```
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModuleRegistrar = void 0;
exports.resolveModuleOrder = resolveModuleOrder;
// ── Module Registrar ────────────────────────────────────────────
/**
 * Builder passed to {@link ZroModule.register} for contributing handlers.
 *
 * Mirrors the `ZroApp` API so modules register in the same way
 * as manual inline handlers.
 */
class ModuleRegistrar {
    /** @internal */ _commands = new Map();
    /** @internal */ _eventHandlers = new Map();
    /** @internal */ _lifecycleHandlers = new Map();
    /** @internal */ _initHooks = [];
    /** @internal */ _destroyHooks = [];
    /** Register a command handler (WS invoke + HTTP API). */
    command(name, handler) {
        this._commands.set(name, handler);
        return this;
    }
    /** Register a WS event handler (fire-and-forget). */
    onEvent(event, handler) {
        this._eventHandlers.set(event, handler);
        return this;
    }
    /** Register a lifecycle handler (`client:connected`, etc.). */
    on(event, handler) {
        this._lifecycleHandlers.set(event, handler);
        return this;
    }
    /** Register an init hook, called after IPC handshake before the main loop. */
    onInit(handler) {
        this._initHooks.push(handler);
        return this;
    }
    /** Register a destroy hook, called during shutdown (reverse init order). */
    onDestroy(handler) {
        this._destroyHooks.push(handler);
        return this;
    }
}
exports.ModuleRegistrar = ModuleRegistrar;
// ── Dependency Resolution ───────────────────────────────────────
/**
 * Resolve module initialization order via topological sort.
 * Returns indices into `modules` in the order they should be initialized.
 *
 * @throws Error on missing dependencies or circular references.
 */
function resolveModuleOrder(modules) {
    const nameToIdx = new Map();
    for (let i = 0; i < modules.length; i++) {
        nameToIdx.set(modules[i].meta.name, i);
    }
    const n = modules.length;
    const inDegree = new Array(n).fill(0);
    const adj = Array.from({ length: n }, () => []);
    for (let i = 0; i < n; i++) {
        const deps = modules[i].meta.dependencies ?? [];
        for (const dep of deps) {
            const depIdx = nameToIdx.get(dep);
            if (depIdx === undefined) {
                throw new Error(`Module '${modules[i].meta.name}' depends on '${dep}' which is not registered`);
            }
            adj[depIdx].push(i);
            inDegree[i]++;
        }
    }
    // Kahn's algorithm
    const queue = [];
    for (let i = 0; i < n; i++) {
        if (inDegree[i] === 0)
            queue.push(i);
    }
    const order = [];
    while (queue.length > 0) {
        const node = queue.shift();
        order.push(node);
        for (const next of adj[node]) {
            inDegree[next]--;
            if (inDegree[next] === 0) {
                queue.push(next);
            }
        }
    }
    if (order.length !== n) {
        throw new Error('Circular dependency detected among modules');
    }
    return order;
}
